File: busing-bridge/src/busing.rb
require_relative "packet"
require "timeout"

require_relative "input_output"
require_relative "kctr"
require_relative "smart_touch_plus"

class Busing
  READ_MEM = 3
  WRITE_MEM = 4
  READ_EEPROM = 5
  READ_DEV_TYPE = 9
  READ_ADDRESS = 7

  DEV_TYPES = {
    23 => "KCTR_KA",
    24 => "2E2S",
    6 => "SMART_TOUCH"
  }

  UNKNOWN_DEVICE_TYPE = "UNKNOWN"
  MAX_DEVICE_ID = 255 # without routing this is the max id

  PACKET_READING_TIMEOUT = 0.1
  WAIT_WHEN_NO_PACKETS = 0.1
  WAITING_TIME_BEFORE_RECONNECT = 2

  class << self
    def connect(host:, port: 12347, max_devices_to_found: 255, log_level: Logger::WARN)
      logger = Logger.new(STDOUT)
      logger.level = log_level
      logger.info("Busing connecting")
      new(host: host, port: port, logger: logger).tap do |instance|
        t0 = Time.now
        instance.discover_devices(max_devices_to_found: max_devices_to_found)
        logger.info("Busing ready in #{Time.now - t0}!")
      end
    end
  end

  def initialize(host:, port:, logger:)
    @host = host
    @port = port
    @device = []
    @socket = nil
    @logger = logger
  end

  def reconnect!
    socket(reconnect: true)
  end

  def socket(reconnect: false)
    raise "Busing not configured" if @host.nil?
    if reconnect
      @socket.close unless @socket.nil?
      @socket = nil
      logger.info("Reconnecting to busing...")
    end
    @socket ||= TCPSocket.new(@host, @port)
  end

  def logger
    @logger
  end

  def next_datagram
    Timeout::timeout(PACKET_READING_TIMEOUT) do
      readed = socket.read(Packet::SIZE)
      return Packet::NIL_PACKET if readed.nil?
      
      readed.bytes
    end
  end

  def next_packet(&block)
    Packet.response(next_datagram)
  rescue Timeout::Error
    block.call if block_given?
    sleep WAIT_WHEN_NO_PACKETS
    :no_packet
  end

  def discover_devices(max_devices_to_found: 255)
    (0..MAX_DEVICE_ID).map do |dir|
      packet = Packet.command(Busing::READ_DEV_TYPE)
      packet.address_to = dir
      packet.data1 = 0
      packet.data2 = 0
      socket.write(packet.as_command.pack("C*"))
    end
  
    @devices = []
    responses = []
    received = 0 
    while received < MAX_DEVICE_ID && responses.count < max_devices_to_found do
      packet = next_packet
      next if packet == :no_packet
      if packet.address_to == 65278
        received += 1
        responses << packet if packet.command == 1
        logger.debug("New device found")
      end
    end
    responses.map do |response|
      device_id = response.address_from
      response = send_command_to_address(Busing::READ_DEV_TYPE, device_id, 0, 0)
      dev_type = DEV_TYPES[response.data1] || UNKNOWN_DEVICE_TYPE
      devices << {
        device_id: device_id,
        node_type: dev_type,
        controller: new_controller_for(dev_type)
      }
    end
    logger.info("#{devices.count} found.")
  end

  def output_state_by(name:)
    # Buscar en outputs normales
    device = devices.find { |device| device[:controller].output_names.include?(name) }
    return device[:controller].output_state_by(name: name) if device
    
    # Buscar en registros
    device = devices.find do |device| 
      device[:controller].respond_to?(:registers_config) && 
      device[:controller].registers_config.any? { |reg| reg["entity_name"] == name }
    end
    return nil if device.nil?
    
    # Para registros, devolver estado por defecto (se actualizará cuando lleguen eventos)
    "OFF"
  end

  def set_state_by(name:, value:)
    # Buscar en outputs normales
    device = devices.find { |device| device[:controller].output_names.include?(name) }
    return device[:controller].set_state_by(name: name, value: value) if device
    
    # Los registros normalmente son de solo lectura (sensores)
    # pero podríamos agregar soporte si es necesario
    logger.warn("Cannot set state for register entity '#{name}' - registers are typically read-only")
    nil
  end

  def input_state_by(name:)
    device = devices.find { |device| device[:controller].input_names.include?(name) }
    return nil if device.nil?
    device[:controller].input_state_by(name: name)
  end

  def configure_device(device_type, outputs:, inputs:, registers:)
    device = devices.find { |d| d[:node_type] == device_type }
    logger.info("Configuring '#{device_type}'")
    logger.debug("Configuration for '#{device[:controller].class.name}': '#{outputs}'")
    device[:controller].output_names = outputs
    device[:controller].input_names = inputs
    if registers
      device[:controller].registers_config = registers.map do |register_config|
        id, entity_name = register_config.split(":")
        { "id" => id, "entity_name" => entity_name}
      end
    end
  end

  def devices
    @devices
  end

  def new_controller_for(type)
    case type
    when "KCTR_KA"
      KCtr.new(busing: self)
    when "2E2S"
      InputOutput.new(busing: self, device_type: "2E2S", type: "2E2S")
    when "SMART_TOUCH"
      SmartTouchPlus.new(busing: self)
    end
  end

  def send_command(cmd, device_type, data1, data2)
    device = devices.find { |d| d[:node_type] == device_type }
    raise "device '#{device_type}' not found!" if device.nil?
    send_command_to_address(cmd, device[:device_id], data1, data2)
  end

  def send_command_to_address(cmd, to, data1, data2)
    packet = Packet.command(cmd)
    packet.address_to = to
    packet.data1 = data1
    packet.data2 = data2
    cmd = packet.as_command
    socket.write(cmd.pack("C*"))
    packet = next_packet
    while packet == :no_packet || packet.address_to != 65278 || packet.address_from != to do
      packet = next_packet
    end
    packet
  end

  def decode(packet)
    info = { action: "noop", packet: packet }

    device = devices.find { |d| d[:device_id] == packet.address_to }

    return info if device.nil? # unknown device

    decoded_packet = device[:controller].decode(packet)

    info.merge(decoded_packet, packet: nil).compact
  end

  def listen(&block)
    loop do
      while packet = next_packet(&method(:not_packet)) do
        if packet != :no_packet && !packet.valid?
          invalid_packet(packet)
          reconnect!
          next
        end

        info = if packet == :no_packet
                 :no_event
               else
                 decode(packet)
               end
          
        block.call(info, packet)
      end
      logger.info("No more packets, reconnecting...")
      sleep(WAITING_TIME_BEFORE_RECONNECT)
      reconnect!    
    end
  end

  def invalid_packet(packet)
    logger.warn("Invalid Packet: #{packet.inspect}")
  end

  def not_packet
    logger.debug("No packets received...")
  end
end

File: busing-bridge-py/src/busing.py
"""TCP connection to the Busing installation: discovery, commands, event stream."""

import logging
import socket
import time
from dataclasses import dataclass, field

from const import DEV_TYPES, READ_DEV_TYPE, UNKNOWN_DEVICE_TYPE
from controllers import BaseController, InputOutput, KCtr, SmartTouchPlus
from packet import Packet


@dataclass
class Device:
    device_id: int
    node_type: str
    controller: BaseController = field(repr=False)


class Busing:
    # Replies from devices are addressed to the TCP gateway at this address
    BRIDGE_ADDRESS = 65278
    MAX_DEVICE_ID = 255  # without routing this is the max id

    PACKET_READ_TIMEOUT = 0.1  # seconds
    RECONNECT_DELAY = 2  # seconds
    DISCOVERY_TIMEOUT = 30  # seconds
    COMMAND_TIMEOUT = 5  # seconds

    @classmethod
    def connect(cls, host, port=12347, max_devices=255, logger=None):
        logger = logger or logging.getLogger("busing")
        logger.info("Busing connecting")
        instance = cls(host=host, port=port, logger=logger)
        started = time.monotonic()
        instance.discover_devices(max_devices=max_devices)
        logger.info("Busing ready in %.1fs!", time.monotonic() - started)
        return instance

    def __init__(self, host, port, logger):
        if not host:
            raise ValueError("busing_host is required")
        self.host = host
        self.port = port
        self.logger = logger
        self.devices = []
        self._sock = None
        self._buffer = b""

    def _socket(self):
        if self._sock is None:
            self.logger.info("Connecting to busing at %s:%s", self.host, self.port)
            self._sock = socket.create_connection((self.host, self.port))
            self._sock.settimeout(self.PACKET_READ_TIMEOUT)
            self._buffer = b""
        return self._sock

    def reconnect(self):
        if self._sock is not None:
            try:
                self._sock.close()
            except OSError:
                pass
            self._sock = None
        self.logger.info("Reconnecting to busing...")
        return self._socket()

    def next_packet(self):
        """Return the next packet, or None if nothing arrived within the read timeout.

        Partial reads are buffered, so a slow sender never desynchronizes the
        9-byte framing.
        """
        sock = self._socket()
        while len(self._buffer) < Packet.SIZE:
            try:
                chunk = sock.recv(4096)
            except socket.timeout:
                return None
            if not chunk:
                raise ConnectionError("Busing closed the connection")
            self._buffer += chunk
        data, self._buffer = self._buffer[: Packet.SIZE], self._buffer[Packet.SIZE:]
        return Packet(data)

    def discover_devices(self, max_devices=255):
        sock = self._socket()
        for address in range(self.MAX_DEVICE_ID + 1):
            packet = Packet.for_command(READ_DEV_TYPE)
            packet.address_to = address
            sock.sendall(packet.as_command())

        addresses = []
        received = 0
        deadline = time.monotonic() + self.DISCOVERY_TIMEOUT
        while received < self.MAX_DEVICE_ID and len(addresses) < max_devices:
            if time.monotonic() > deadline:
                self.logger.warning(
                    "Device discovery timed out after %ss with %s device(s) found "
                    "(check 'busing_devices_installed')",
                    self.DISCOVERY_TIMEOUT,
                    len(addresses),
                )
                break
            packet = self.next_packet()
            if packet is None:
                continue
            if packet.address_to != self.BRIDGE_ADDRESS:
                continue
            received += 1
            if packet.command == 1:
                addresses.append(packet.address_from)
                self.logger.debug("New device found at address %s", packet.address_from)

        self.devices = []
        for device_id in addresses:
            response = self.send_command_to_address(READ_DEV_TYPE, device_id, 0, 0)
            node_type = DEV_TYPES.get(response.data1, UNKNOWN_DEVICE_TYPE)
            self.devices.append(
                Device(
                    device_id=device_id,
                    node_type=node_type,
                    controller=self._new_controller(node_type),
                )
            )
        self.logger.info("%s device(s) found.", len(self.devices))

    def _new_controller(self, node_type):
        if node_type == "KCTR_KA":
            return KCtr(busing=self)
        if node_type == "2E2S":
            return InputOutput(busing=self, device_type="2E2S", module_type="2E2S")
        if node_type == "SMART_TOUCH":
            return SmartTouchPlus(busing=self)
        return BaseController(busing=self)

    def configure_device(self, device_type, outputs=None, inputs=None, registers=None):
        device = next((d for d in self.devices if d.node_type == device_type), None)
        if device is None:
            self.logger.warning("Cannot configure '%s': no such device discovered", device_type)
            return
        self.logger.info("Configuring '%s'", device_type)
        if outputs is not None:
            device.controller.output_names = outputs
        if inputs is not None:
            device.controller.input_names = inputs
        if registers:
            device.controller.registers_config = [
                dict(zip(("id", "entity_name"), register.split(":", 1)))
                for register in registers
            ]

    def output_state_by(self, name):
        device = next(
            (d for d in self.devices if name in d.controller.output_names), None
        )
        if device is not None:
            return device.controller.output_state_by(name)

        # Register entities report a default state until events arrive
        device = next(
            (
                d
                for d in self.devices
                if any(reg["entity_name"] == name for reg in d.controller.registers_config)
            ),
            None,
        )
        if device is None:
            return None
        return "OFF"

    def set_state_by(self, name, value):
        device = next(
            (d for d in self.devices if name in d.controller.output_names), None
        )
        if device is not None:
            return device.controller.set_state_by(name, value)
        self.logger.warning(
            "Cannot set state for register entity '%s' - registers are typically read-only",
            name,
        )
        return None

    def input_state_by(self, name):
        device = next(
            (d for d in self.devices if name in d.controller.input_names), None
        )
        if device is None:
            return None
        return device.controller.input_state_by(name)

    def send_command(self, cmd, device_type, data1, data2):
        device = next((d for d in self.devices if d.node_type == device_type), None)
        if device is None:
            raise LookupError(f"device '{device_type}' not found!")
        return self.send_command_to_address(cmd, device.device_id, data1, data2)

    def send_command_to_address(self, cmd, to, data1, data2):
        packet = Packet.for_command(cmd)
        packet.address_to = to
        packet.data1 = data1
        packet.data2 = data2
        self._socket().sendall(packet.as_command())

        deadline = time.monotonic() + self.COMMAND_TIMEOUT
        while time.monotonic() < deadline:
            response = self.next_packet()
            if response is None:
                continue
            if response.address_to == self.BRIDGE_ADDRESS and response.address_from == to:
                return response
        raise TimeoutError(f"No response from device {to} to command {cmd}")

    def decode(self, packet):
        info = {"action": "noop"}
        device = next(
            (d for d in self.devices if d.device_id == packet.address_to), None
        )
        if device is None:  # unknown device
            return info
        info.update(device.controller.decode(packet))
        return info

    def listen(self):
        """Yield (event, packet) tuples forever, reconnecting as needed.

        Yields (None, None) when no packet arrived within the read timeout so
        the caller can run periodic work (resync, MQTT commands).
        """
        while True:
            try:
                packet = self.next_packet()
            except (ConnectionError, OSError) as err:
                self.logger.warning("Busing connection lost (%s), reconnecting...", err)
                time.sleep(self.RECONNECT_DELAY)
                try:
                    self.reconnect()
                except OSError as reconnect_err:
                    self.logger.error("Reconnect failed: %s", reconnect_err)
                continue

            if packet is None:
                yield None, None
                continue

            if not packet.is_valid():
                self.logger.warning("Invalid packet: %s", packet.describe())
                self.reconnect()
                continue

            yield self.decode(packet), packet
